### PR TITLE
Use the improved submodule handling

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
r? @alexcrichton 

That was a crap...
```
Updating submodules
Traceback (most recent call last):
  File "./x.py", line 20, in <module>
    bootstrap.main()
  File "/home/ishitatsuyuki/Documents/rust/src/bootstrap/bootstrap.py", line 684, in main
    bootstrap()
  File "/home/ishitatsuyuki/Documents/rust/src/bootstrap/bootstrap.py", line 662, in bootstrap
    rb.update_submodules()
  File "/home/ishitatsuyuki/Documents/rust/src/bootstrap/bootstrap.py", line 566, in update_submodules
    path = line[1:].split(' ')[1]
TypeError: a bytes-like object is required, not 'str'
```

Maybe we need to confirm the compatibility of git options, such as `git config` or `git -C` (I believe they existed long before, though). This is tested locally.

### CODE REVIEW

**Comment by michael-zapata** [comment]
Just a quick feedback on that: `git -C` isn't *that* old, RHEL6 doesn't handle it for example.

**Comment by alexcrichton** [comment]
r? @aidanhs 

Mind taking a look at this? I think you know more about submodules than I

**Comment by kennytm** [comment]
@ishitatsuyuki The Python 2/3 thing is the same as #42085? (fixed in #42089)

**Comment by ishitatsuyuki** [comment]
I have rebased the things. This should supersede #42089, if it was possible to gain approval.

... sorry for the redundant formatting, I couldn't keep up without my auto formatter.

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -545,51 +557,30 @@ def build_triple(self):
...
-                self.run(["git", "submodule", "init", path], cwd=self.rust_root)
-                self.run(["git", "submodule", "update", path], cwd=self.rust_root)
-            else:
-                raise ValueError('unknown submodule status: ' + status)
+        self.run(["git", "submodule", "-q", "sync"], cwd=self.rust_root)
+        # FIXME: nobody does, but this won't work well with whitespace in
+        # submodule path
+        submodules = [s.split()[1] for s in subprocess.check_output(
+            ["git", "config", "--file", os.path.join(
+                self.rust_root, ".gitmodules"), "--get-regexp", "path"]).splitlines()]
```
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: Unfortunately we probably do actually need to do the defaultencoding dance even here, doesn't windows sometimes use two-byte unicode encodings? Which would mean checking for `b"llvm"` wouldn't work.

Comment (ishitatsuyuki) [inline src/bootstrap/bootstrap.py]: I don't think so. The command line uses the current code page, which IIRC is ASCII compat. Minding trying out that?

**Comment by ishitatsuyuki** [comment]
Rebase made with the `theirs` strategy option.

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -546,51 +557,30 @@ def build_triple(self):
...
+        submodules = [module for module in submodules
+                      if not ((module.endswith(b"llvm") and
+                              (self.get_toml('llvm-config') or self.get_mk('CFG_LLVM_ROOT'))) or
+                      (module.endswith(b"jemalloc") and
```
Comment (Mark-Simulacrum) [inline src/bootstrap/bootstrap.py]: I'm fairly certain that using `b"..."` will break Python 3 compatibility, see #42089.

Comment (ishitatsuyuki) [inline src/bootstrap/bootstrap.py]: ... what if I tell you that I happily ran it on my Arch laptop which has Py3?

Comment (Mark-Simulacrum) [inline src/bootstrap/bootstrap.py]: Hm, odd. Not sure why it was changed in that PR then. Never mind me.

Comment (ishitatsuyuki) [inline src/bootstrap/bootstrap.py]: It seems that the original PR decoded the pipe output which is in system default encoding, and can vary on Windows. However, I'm quite sure that all code pages are ASCII compat, which doesn't pose a problem if you're dealing with pure English paths. (want some emoji there? :P)

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: You are indeed correct.

That said, I feel like being consistent with the rest of the file (which uses defaultencoding) is probably worthwhile. Indeed, two reviewers of this PR have already been confused by imagined possible problems!

File: src/bootstrap/bootstrap.py (hunk 2 above)
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: The original PR took didn't decode at all, which was the problem - it took a bytestring of submodules and tried split it with a string, which isn't allowed. The followup PR made sure the bytestring was converted to a string like all the other outputs from commands in the file, so now using bytestrings would be incorrect. This PR goes back to not decoding command output, so bytestrings are correct again.

**Comment by aidanhs** [comment]
Tested on Centos 6, r=me after moving from bytestrings to defaultencoding for consistency (if you happen to undo the autoformat diff noise at the same time then that'd also be fantastic).

**Comment by ishitatsuyuki** [comment]
@aidanhs I'm not sure if `getdefaultencoding()` is a good idea. It's almost constant ('ascii' for 2, 'utf8' for 3), and isn't so useful for Windows.

**Comment by ishitatsuyuki** [comment]
Roughly tested on both Arch and CentOS 6.

**Comment by aidanhs** [comment]
Great! Thanks very much for bearing with me @ishitatsuyuki :)

@bors r+

**Comment by ishitatsuyuki** [comment]
Rebased. Didn't tested on CentOS yet, but I believe it should be fine. Works fine on Arch with both Python 2/3.

**Comment by ishitatsuyuki** [comment]
Review please. Basically I think this is tested and reviewed enough and it's fine to r+ as I only did a rebase.

**Pull request merged**